### PR TITLE
[SPARK-17741][SQL] Grammar to parse top level and nested data fields separately

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Currently we use the same rule to parse top level and nested data fields. For example:

```
create table tbl_x(
  id bigint,
  nested struct<col1:string,col2:string>
)
```

Shows both syntaxes. In this PR we split this rule in a top-level and nested rule.

Before this PR,

```
sql("CREATE TABLE my_tab(column1: INT)")
```

works fine.
After this PR, it will throw a `ParseException`:

```
scala> sql("CREATE TABLE my_tab(column1: INT)")
org.apache.spark.sql.catalyst.parser.ParseException:
no viable alternative at input 'CREATE TABLE my_tab(column1:'(line 1, pos 27)
```
## How was this patch tested?

Add new testcases in `SparkSqlParserSuite`.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala [hunk 1]
```diff
@@ -67,9 +86,331 @@ class SparkSqlParserSuite extends PlanTest {
...
+        storage = storage,
+        schema = schema,
+        provider = provider,
+        partitionColumnNames = partitionColumnNames,
+        comment = comment
+      ), mode, query
+    )
+  }
+
+  test("create table using - createTableHeader") {
```
Comment (hvanhovell) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala]: Most of these unit tests are not related to this change, lets move these into a separate PR. Just put a couple of cases (at least one positive and one negative case) in here testing your change.

**Review by hvanhovell — CHANGES_REQUESTED** [review]
This looks pretty good. Could you add a few more tests to DataTypeParseSuite, and tone down the testing in SparkSqlParserSuite?

**Comment by jiangxb1987** [comment]
@hvanhovell Thank you for your suggestion! I've addressed your comment and amended the testcases. Thanks!

**Comment by jiangxb1987** [comment]
retest this please

**Comment by jiangxb1987** [comment]
@hvanhovell Could you please look at this? Thank you!

**Comment by hvanhovell** [comment]
LGTM - merging to master. Thanks!

**Comment by HyukjinKwon** [comment]
Hi @hvanhovell,  it seems it fails to build now. I guess you might already know but I am just leaving a comment here just in case to let you know.

```
[error] /home/jenkins/workspace/spark-master-compile-maven-scala-2.10/sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala:126: not enough arguments for method apply: (tableIdent: org.apache.spark.sql.catalyst.TableIdentifier, userSpecifiedSchema: Option[org.apache.spark.sql.types.StructType], replace: Boolean, global: Boolean, provider: String, options: Map[String,String])org.apache.spark.sql.execution.datasources.CreateTempViewUsing in object CreateTempViewUsing.
[error] Unspecified value parameter options.
[error]     CreateTempViewUsing(TableIdentifier(table, database), schema, replace, provider, options)
[error]                        ^
[error] one error found
[error] Compile failed at Oct 10, 2016 1:01:12 AM [23.927s]
```

**Comment by jiangxb1987** [comment]
@HyukjinKwon I'll fix that in a follow up. Thank you for telling!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala [hunk 2]
```diff
@@ -68,9 +87,134 @@ class SparkSqlParserSuite extends PlanTest {
...
+        storage = storage,
+        schema = schema,
+        provider = provider,
+        partitionColumnNames = partitionColumnNames,
+        bucketSpec = bucketSpec
+      ), mode, query
+    )
+  }
+
+  private def createTempViewUsing(
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlParserSuite.scala]: yeah, this newly added function seems unused.
